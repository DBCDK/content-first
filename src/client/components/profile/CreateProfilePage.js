import React from 'react';
import {connect} from 'react-redux';

import ImageUpload from '../general/ImageUpload.component';
import UserProfileForm from './UserProfileForm.component';
import Spinner from '../general/Spinner.component';
import Link from '../general/Link.component';
import LogoutLink from '../general/Logout.component';
import {
  ADD_PROFILE_IMAGE,
  SAVE_USER_PROFILE,
  DELETE_USER_PROFILE
} from '../../redux/user.reducer';

export class CreateProfilePage extends React.Component {
  render() {
    const isDeleting = this.props.isDeleting || false;

    if (isDeleting) {
      return <Spinner className="spinner-deleteProfile" size="36px" />;
    }

    return (
      <div className="container">
        <div className="profile-page tl">
          <h1 className="headline mt2">{this.props.title}</h1>
          <div className="profile-form mb3">
            {this.props.isLoading ? (
              <div className="mb2 tc">
                <Spinner size="36px" />
              </div>
            ) : (
              <div className="row clearfix m-0">
                <div className="col-3">
                  <ImageUpload
                    error={this.props.imageError}
                    style={{borderRadius: '50%'}}
                    loading={this.props.imageIsLoading}
                    previewImage={
                      this.props.profileImageId
                        ? `/v1/image/${this.props.profileImageId}/150/150`
                        : null
                    }
                    buttonText={
                      this.props.profileImageId
                        ? 'Skift profilbillede'
                        : 'Upload profilbillede'
                    }
                    onFile={this.props.addImage}
                  />
                </div>
                <div className="col-9 mb2">
                  <p className="mb2">
                    <b>Indtast dine personlige oplysninger:</b>
                  </p>
                  <div className="mb2">
                    <UserProfileForm
                      imageId={this.props.profileImageId}
                      name={this.props.name}
                      acceptedTerms={this.props.acceptedTerms}
                      library={this.props.agencyName || <Spinner size="12px" />}
                      updateProfile={this.props.saveUser}
                      error={this.props.error}
                      isSaving={this.props.isSaving}
                      editMode={this.props.editMode}
                    />
                  </div>
                  {this.props.editMode ? (
                    <React.Fragment>
                      <Link
                        href="#!"
                        className="text-danger profile-delete"
                        onClick={() => this.props.confirmDeleteModal()}
                      >
                        Slet profil
                      </Link>
                      {' | '}
                      <Link href="/profile">Fortryd redigéring</Link>
                    </React.Fragment>
                  ) : (
                    <LogoutLink>
                      Jeg ønsker alligevel ikke at oprette en profil
                    </LogoutLink>
                  )}
                </div>
              </div>
            )}
          </div>
          <div id="terms" className="terms">
            Her skal stå regler for anvendelse af læsekompasset...
          </div>
        </div>
      </div>
    );
  }
}

export const mapStateToProps = state => ({
  name: state.userReducer.name,
  acceptedTerms: state.userReducer.acceptedTerms,
  isLoading: state.userReducer.isLoading,
  isSaving: state.userReducer.isSaving,
  isDeleting: state.userReducer.isDeleting,
  profileImageId: state.userReducer.tempImageId || state.userReducer.image,
  imageIsLoading: state.userReducer.profileImageIsLoading,
  error: state.userReducer.saveUserError,
  imageError: state.userReducer.imageError,
  agencyName: state.userReducer.agencyName
});
export const mapDispatchToProps = dispatch => {
  const closeModal = (type, modal) => dispatch({type, modal});
  const deleteUser = type => dispatch({type});

  return {
    addImage: image => dispatch({type: ADD_PROFILE_IMAGE, image}),
    saveUser: user => dispatch({type: SAVE_USER_PROFILE, user}),
    closeModal,
    confirmDeleteModal: () => {
      dispatch({
        type: 'OPEN_MODAL',
        modal: 'confirm',
        context: {
          title: 'Er du sikker på du vil slette din profil?',
          reason:
            'Du er ved at slette din profil og alt data som er tilknyttet den, Er du sikke rpå du vil fortsætte?.',
          confirmText: 'Slet min profil',
          onConfirm: () => {
            closeModal('CLOSE_MODAL', 'confirm');
            deleteUser(DELETE_USER_PROFILE);
          },
          onCancel: () => closeModal('CLOSE_MODAL', 'confirm')
        }
      });
    }
  };
};
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CreateProfilePage);
