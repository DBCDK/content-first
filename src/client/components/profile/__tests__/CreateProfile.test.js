/* eslint-disable no-undefined */
import React from 'react';
import renderer from 'react-test-renderer';
import {mount} from 'enzyme';

import {CreateProfilePage} from '../CreateProfilePage';
const image =
  'data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==';
jest.mock('../../general/Spinner.component', () => 'spinner');

describe('Image Upload', () => {
  describe('render function', () => {
    it('renders an image upload form', () => {
      const tree = renderer.create(<CreateProfilePage />).toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('Add image', done => {
      const onFile = jest.fn();
      const tree = mount(<CreateProfilePage addImage={onFile} />);
      tree
        .find('input')
        .first()
        .simulate('change', {target: {files: [image]}});
      tree.update();
      expect(tree).toMatchSnapshot();
      expect(onFile.mock.calls[0][0]).toEqual(image);
      done();
    });
  });
});
