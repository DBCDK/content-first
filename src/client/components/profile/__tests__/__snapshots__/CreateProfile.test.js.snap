// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image Upload render function Add image 1`] = `
<CreateProfilePage
  addImage={[Function]}
>
  <div
    className="profile__edit-content"
  >
    <Head
      canonical="/profil/rediger"
      robots="noindex"
      title="Min profil"
    >
      <HelmetWrapper
        defer={true}
        encodeSpecialCharacters={true}
      >
        <SideEffect(NullComponent)
          defer={true}
          encodeSpecialCharacters={true}
          link={
            Array [
              Object {
                "href": "https://laesekompas.dk/profil/rediger",
                "rel": "canonical",
              },
            ]
          }
          meta={
            Array [
              Object {
                "content": "noindex",
                "name": "robots",
              },
              Object {
                "content": "På Læsekompasset kan du gå på opdagelse i skønlitteraturen, få personlige anbefalinger og dele dine oplevelser med andre.",
                "name": "description",
              },
              Object {
                "content": "Min profil",
                "name": "og:title",
              },
              Object {
                "content": "På Læsekompasset kan du gå på opdagelse i skønlitteraturen, få personlige anbefalinger og dele dine oplevelser med andre.",
                "name": "og:description",
              },
              Object {
                "content": "https://laesekompas.dk/",
                "property": "og:url",
              },
              Object {
                "content": "Læsekompas",
                "property": "og:site_name",
              },
              Object {
                "content": "website",
                "property": "og:type",
              },
              Object {
                "content": "http://laesekompas.dk/img/open-graph/hero-01.jpg",
                "property": "og:image",
              },
              Object {
                "content": "1200",
                "property": "og:image:width",
              },
              Object {
                "content": "675",
                "property": "og:image:height",
              },
            ]
          }
          title="Min profil | Læsekompas"
          titleAttributes={Object {}}
        >
          <NullComponent
            defer={true}
            encodeSpecialCharacters={true}
            link={
              Array [
                Object {
                  "href": "https://laesekompas.dk/profil/rediger",
                  "rel": "canonical",
                },
              ]
            }
            meta={
              Array [
                Object {
                  "content": "noindex",
                  "name": "robots",
                },
                Object {
                  "content": "På Læsekompasset kan du gå på opdagelse i skønlitteraturen, få personlige anbefalinger og dele dine oplevelser med andre.",
                  "name": "description",
                },
                Object {
                  "content": "Min profil",
                  "name": "og:title",
                },
                Object {
                  "content": "På Læsekompasset kan du gå på opdagelse i skønlitteraturen, få personlige anbefalinger og dele dine oplevelser med andre.",
                  "name": "og:description",
                },
                Object {
                  "content": "https://laesekompas.dk/",
                  "property": "og:url",
                },
                Object {
                  "content": "Læsekompas",
                  "property": "og:site_name",
                },
                Object {
                  "content": "website",
                  "property": "og:type",
                },
                Object {
                  "content": "http://laesekompas.dk/img/open-graph/hero-01.jpg",
                  "property": "og:image",
                },
                Object {
                  "content": "1200",
                  "property": "og:image:width",
                },
                Object {
                  "content": "675",
                  "property": "og:image:height",
                },
              ]
            }
            title="Min profil | Læsekompas"
            titleAttributes={Object {}}
          />
        </SideEffect(NullComponent)>
      </HelmetWrapper>
    </Head>
    <div
      className="profile__edit-titlezone"
    >
      <h1
        className="profile__edit-title"
      >
        <T
          component="profile"
          name="editProfile"
        >
          Rediger profil
        </T>
      </h1>
    </div>
    <div
      className="profile__edit-page-position"
    >
      <div
        className="profile__edit-top-position"
      >
        <div
          className="profile__edit-image-zone"
        >
          <div
            className="profile__edit-image-upload"
          >
            <ProfileUploadImage
              activateSaveButton={[Function]}
              baseImage="/img/general/user-placeholder-thumbnail.svg"
              onFile={[Function]}
              thumbnailImageHover="/img/general/user-placeholder-thumbnail-hover.svg"
            >
              <div
                className="profile-picture"
              >
                <div
                  className="image-upload"
                >
                  <div
                    className="image-upload-loader"
                  >
                    <UploadButton
                      baseImage="/img/general/user-placeholder-thumbnail.svg"
                      readFiles={[Function]}
                      thumbnailImageHover="/img/general/user-placeholder-thumbnail-hover.svg"
                    >
                      <label>
                        <HoverImage
                          mouseOver="/img/general/user-placeholder-thumbnail-hover.svg"
                          src="/img/general/user-placeholder-thumbnail.svg"
                        >
                          <img
                            alt=""
                            className="image-upload-image"
                            onMouseOut={[Function]}
                            onMouseOver={[Function]}
                            src="/img/general/user-placeholder-thumbnail.svg"
                          />
                        </HoverImage>
                        <input
                          accept="image/png, image/jpeg"
                          className="droppable-image-field--file-input"
                          onChange={[Function]}
                          type="file"
                        />
                        <div
                          className="profile-image-text"
                        >
                          <T
                            component="profile"
                            name="uploadProfilePicture"
                          >
                            Vælg billede
                          </T>
                        </div>
                      </label>
                    </UploadButton>
                  </div>
                  <div
                    className="profile__image-errors"
                  >
                    <Error />
                  </div>
                </div>
              </div>
            </ProfileUploadImage>
          </div>
        </div>
        <div
          className="profile__edit-textfield-zone"
        >
          <EditProfileForm
            activateSaveButton={[Function]}
            confirmDelete={[Function]}
            enableButton={true}
          >
            <form
              onSubmit={[Function]}
            >
              <div
                className="input-group mb2   has-feedback"
              >
                <div
                  className="profile__edit-username-zone"
                >
                  <div
                    className="profile__edit-inputNameTitle"
                  >
                    <T
                      component="profile"
                      name="usernameTitle"
                    >
                      Brugernavn
                    </T>
                  </div>
                  <div
                    className="profile__edit-textfield"
                  >
                    <input
                      className="profile__mobileinputfield   has-feedback"
                      data-cy="user-form-name"
                      name="name"
                      onChange={[Function]}
                      placeholder="Skriv brugernavn"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
              <div
                className="profile__edit-libraryinfo"
              >
                <p
                  className="mb6"
                >
                  <T
                    component="login"
                    name="signedInby"
                  >
                    Du er logget på via
                  </T>
                  <input
                    className="profile__edit-libraryname"
                    disabled={true}
                    name="name"
                    type="text"
                    value=""
                  />
                </p>
              </div>
              <div
                className="profile__edit-link"
              >
                <a
                  href="/om"
                >
                  <T
                    component="profile"
                    name="termsAndConditions"
                  >
                    Se regler og vilkår for Læsekompasset
                  </T>
                </a>
              </div>
              <div
                className="profile__edit-link"
              >
                <a
                  href="#!"
                  onClick={[Function]}
                >
                  <T
                    component="profile"
                    name="deleteProfile"
                  >
                    Slet profil
                  </T>
                </a>
              </div>
              <div
                className="profile__edit-bottom-position"
              >
                <Toolbar>
                  <div
                    className="Toolbar "
                  >
                    <div
                      className="Toolbar__left"
                    />
                    <div
                      className="Toolbar__center"
                    />
                    <div
                      className="Toolbar__right"
                    >
                      <Button
                        data-cy="user-form-cancel"
                        key="0"
                        onClick={[Function]}
                        size="medium"
                        style={
                          Object {
                            "backgroundColor": "var(--porcelain)",
                          }
                        }
                        type="quaternary"
                      >
                        <button
                          className="
                              btn Button Button__medium Button__quaternary "
                          data-cy="user-form-cancel"
                          disabled={false}
                          onClick={[Function]}
                          style={
                            Object {
                              "backgroundColor": "var(--porcelain)",
                            }
                          }
                        >
                          <span>
                            <T
                              component="general"
                              name="cancel"
                            >
                              Fortryd
                            </T>
                          </span>
                        </button>
                      </Button>
                      <Button
                        data-cy="user-form-submit"
                        disabled={false}
                        key="1"
                        size="medium"
                        type="quaternary"
                      >
                        <button
                          className="
                              btn Button Button__medium Button__quaternary "
                          data-cy="user-form-submit"
                          disabled={false}
                        >
                          <span>
                            <T
                              component="profile"
                              name="saveChanges"
                            >
                              Gem ændringer
                            </T>
                          </span>
                        </button>
                      </Button>
                    </div>
                  </div>
                </Toolbar>
              </div>
            </form>
          </EditProfileForm>
        </div>
      </div>
    </div>
  </div>
</CreateProfilePage>
`;

exports[`Image Upload render function renders an image upload form 1`] = `
<div
  className="profile__edit-content"
>
  <div
    className="profile__edit-titlezone"
  >
    <h1
      className="profile__edit-title"
    >
      Rediger profil
    </h1>
  </div>
  <div
    className="profile__edit-page-position"
  >
    <div
      className="profile__edit-top-position"
    >
      <div
        className="profile__edit-image-zone"
      >
        <div
          className="profile__edit-image-upload"
        >
          <div
            className="profile-picture"
          >
            <div
              className="image-upload"
            >
              <div
                className="image-upload-loader"
              >
                <label>
                  <img
                    alt=""
                    className="image-upload-image"
                    onMouseOut={[Function]}
                    onMouseOver={[Function]}
                    src="/img/general/user-placeholder-thumbnail.svg"
                  />
                  <input
                    accept="image/png, image/jpeg"
                    className="droppable-image-field--file-input"
                    name={undefined}
                    onChange={[Function]}
                    type="file"
                  />
                  <div
                    className="profile-image-text"
                  >
                    Vælg billede
                  </div>
                </label>
              </div>
              <div
                className="profile__image-errors"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        className="profile__edit-textfield-zone"
      >
        <form
          onSubmit={[Function]}
        >
          <div
            className="input-group mb2   has-feedback"
          >
            <div
              className="profile__edit-username-zone"
            >
              <div
                className="profile__edit-inputNameTitle"
              >
                Brugernavn
              </div>
              <div
                className="profile__edit-textfield"
              >
                <input
                  className="profile__mobileinputfield   has-feedback"
                  data-cy="user-form-name"
                  name="name"
                  onChange={[Function]}
                  placeholder="Skriv brugernavn"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            className="profile__edit-libraryinfo"
          >
            <p
              className="mb6"
            >
              Du er logget på via
              <input
                className="profile__edit-libraryname"
                disabled={true}
                name="name"
                type="text"
                value=""
              />
            </p>
          </div>
          <div
            className="profile__edit-link"
          >
            <a
              href="/om"
            >
              Se regler og vilkår for Læsekompasset
            </a>
          </div>
          <div
            className="profile__edit-link"
          >
            <a
              href="#!"
              onClick={[Function]}
            >
              Slet profil
            </a>
          </div>
          <div
            className="profile__edit-bottom-position"
          >
            <div
              className="Toolbar "
            >
              <div
                className="Toolbar__left"
              />
              <div
                className="Toolbar__center"
              />
              <div
                className="Toolbar__right"
              >
                <button
                  className="
                      btn Button Button__medium Button__quaternary "
                  data-cy="user-form-cancel"
                  disabled={false}
                  onClick={[Function]}
                  style={
                    Object {
                      "backgroundColor": "var(--porcelain)",
                    }
                  }
                >
                  
                  <span>
                    Fortryd
                  </span>
                  
                </button>
                <button
                  className="
                      btn Button Button__medium Button__quaternary "
                  data-cy="user-form-submit"
                  disabled={true}
                >
                  
                  <span>
                    Gem ændringer
                  </span>
                  
                </button>
              </div>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
